libbeat/licenser/elastic_fetcher_test.go

```diff
@@ -0,0 +1,137 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package licenser
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	uuid "github.com/satori/go.uuid"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/elastic/beats/libbeat/outputs/elasticsearch"
+)
+
+func newServerClientPair(t *testing.T, handler http.HandlerFunc) (*httptest.Server, *elasticsearch.Client) {
+	mux := http.NewServeMux()
+	mux.Handle("/_xpack/", http.HandlerFunc(handler))
+
+	server := httptest.NewServer(mux)
+
+	client, err := elasticsearch.NewClient(elasticsearch.ClientSettings{URL: server.URL}, nil)
+	if err != nil {
+		t.Fatalf("could not create the elasticsearch client, error: %s", err)
+	}
+
+	return server, client
+}
+
+func TestParseJSON(t *testing.T) {
+	t.Run("OSS release of Elasticsearch", func(t *testing.T) {
+		h := func(w http.ResponseWriter, r *http.Request) {
+			http.Error(w, "Method Not Allowed", 405)
+		}
+		s, c := newServerClientPair(t, h)
+		defer s.Close()
+		defer c.Close()
+
+		fetcher := NewElasticFetcher(c)
+		oss, err := fetcher.Fetch()
+		if assert.NoError(t, err) {
+			return
+		}
+
+		assert.Equal(t, OSSLicense, oss)
+	})
+
+	t.Run("malformed JSON", func(t *testing.T) {
+		h := func(w http.ResponseWriter, r *http.Request) {
+			w.Write([]byte("hello bad JSON"))
+		}
+		s, c := newServerClientPair(t, h)
+		defer s.Close()
+		defer c.Close()
+
+		fetcher := NewElasticFetcher(c)
+		_, err := fetcher.Fetch()
+		assert.Error(t, err)
+	})
+
+	t.Run("any error from the server", func(t *testing.T) {
+		h := func(w http.ResponseWriter, r *http.Request) {
+			http.Error(w, "Unauthorized", 401)
+		}
+		s, c := newServerClientPair(t, h)
+		defer s.Close()
+		defer c.Close()
+
+		fetcher := NewElasticFetcher(c)
+		_, err := fetcher.Fetch()
+		assert.Error(t, err)
+	})
+
+	t.Run("200 response", func(t *testing.T) {
+		h := func(w http.ResponseWriter, r *http.Request) {
+			json, err := ioutil.ReadFile("data/xpack.json")
```

Comment:
Could we make this `data/xpack-6.4.1.json` (or whatever version it is) and then run a glob here? This makes it easy to add more test files for different versions in the future.

It allows us to also add files from older ES versions.

Reply:
++ love it.